### PR TITLE
[WOR-424] Validate Azure resource providers

### PR DESCRIPTION
**Pull request by aherbst-broad**
## Context
We need to check that needed Azure [resource providers](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-providers-and-types) are registered in a subscription before creating a billing profile, otherwise Terra will be non-functional. 

## This PR
* Adds a check in the verification of the MRG step during profile creation that ensures all required providers are registered. If any are missing, we fail out with a 409.
   * We talked about using 424, but it looks like that is a WebDAV specific error code according to MDN.
* Adds the required providers to the application.yaml configuration.

### CODE REVIEW

File: service/src/main/java/bio/terra/profile/service/profile/flight/create/CreateProfileVerifyDeployedApplicationStep.java [hunk 1]
```diff
@@ -39,12 +46,27 @@ public StepResult doStep(FlightContext context) {
       throw new InaccessibleApplicationDeploymentException(
           String.format(
               "The user [%s] needs access to deployed application [%s] to perform the requested operation",
-              user.getEmail(), profile.managedResourceGroupId().get()));
+              user.getEmail(), profile.getRequiredManagedResourceGroupId()));
     }
 
+    checkRegisteredProviders();
```
Comment (aherbst-broad) [inline service/src/main/java/bio/terra/profile/service/profile/flight/create/CreateProfileVerifyDeployedApplicationStep.java:52]: Not doing this in the `try/catch` above as this is not a retriable check.

**Review by blakery — APPROVED** [review]

File: service/src/main/java/bio/terra/profile/service/azure/AzureService.java [hunk 2]
```diff
@@ -54,6 +61,26 @@ public List<AzureManagedAppModel> getAuthorizedManagedAppDeployments(
         .toList();
   }
 
+  /**
+   * Gets the resource providers in a subscription that are in either the "Registered" or
+   * "Registering" state.
```
Comment (cahrens) [inline service/src/main/java/bio/terra/profile/service/azure/AzureService.java:66]: Out of curiosity, what other states could they be in?

File: service/src/main/java/bio/terra/profile/service/profile/exception/MissingRequiredProvidersException.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package bio.terra.profile.service.profile.exception;
+
+import bio.terra.common.exception.ErrorReportException;
+import java.util.List;
+import org.springframework.http.HttpStatus;
+
+public class MissingRequiredProvidersException extends ErrorReportException {
```
Comment (cahrens) [inline service/src/main/java/bio/terra/profile/service/profile/exception/MissingRequiredProvidersException.java:7]: Should you update the Swagger to document this possible exception from the create method?

File: service/src/main/java/bio/terra/profile/service/azure/AzureService.java [hunk 4]
```diff
@@ -54,6 +61,26 @@ public List<AzureManagedAppModel> getAuthorizedManagedAppDeployments(
...
+   */
+  public Set<String> getRegisteredProvidersForSubscription(UUID tenantId, UUID subscriptionId) {
+    var resourceManager = crlService.getResourceManager(tenantId, subscriptionId);
+    var providers = resourceManager.providers();
+    return providers.list().stream()
+        .filter(
+            provider ->
+                provider.registrationState().equals("Registered")
+                    || provider.registrationState().equals("Registering"))
+        .map(Provider::namespace)
```
Comment (cahrens) [inline service/src/main/java/bio/terra/profile/service/azure/AzureService.java:80]: What does line 80 do? Convert from a `Provider` to just the string `namespace`? I'm wondering if we should return the full `Provider` instead.

Comment (cahrens) [reply service/src/main/java/bio/terra/profile/service/azure/AzureService.java:80]: Or update the documentation to make it clear that it returns the provider namespaces.

File: service/src/main/java/bio/terra/profile/service/azure/AzureService.java (hunk 2 above)
Comment (aherbst-broad) [reply service/src/main/java/bio/terra/profile/service/azure/AzureService.java:66]: NotRegistered

Comment (cahrens) [reply service/src/main/java/bio/terra/profile/service/azure/AzureService.java:66]: Ah, I see. I was imagining it only returned things that they had registered at some point, but possibly things could be in an error state.

File: service/src/main/java/bio/terra/profile/service/azure/AzureService.java (hunk 4 above)
Comment (aherbst-broad) [reply service/src/main/java/bio/terra/profile/service/azure/AzureService.java:80]: Returning the namespace should be sufficient unless there is a compelling need for consumers of this method to have the full provider. I'll update the doc.

File: service/src/main/java/bio/terra/profile/service/profile/exception/MissingRequiredProvidersException.java (hunk 3 above)
Comment (aherbst-broad) [reply service/src/main/java/bio/terra/profile/service/profile/exception/MissingRequiredProvidersException.java:7]: Done, see https://github.com/DataBiosphere/terra-billing-profile-manager/blob/45e2e4491708777d063fdbab60459b013909869d/service/src/main/resources/api/service_openapi.yaml#L65-L71

**Review by cahrens — APPROVED** [review]

**Pull request merged**